### ISSUE TITLE
Refactor UI components

### ISSUE THREAD

**Issue by nmorduch** [wagtail/wagtail#5015]
- [ ] breadcrumb
- [ ] button
- [ ] browser-message
- [ ] chooser
- [ ] dropdown.legacy
- [ ] dropdown
- [ ] footer
- [ ] forms
- [ ] grid.legacy
- [ ] header
- [ ] help-block
- [ ] human-readable-date
- [ ] icons
- [ ] indicator
- [ ] link.legacy
- [ ] listing
- [ ] loading-mask
- [ ] logo
- [ ] main-nav
- [ ] media-placeholder
- [ ] messages.capability
- [ ] messages
- [ ] messages.status
- [ ] modals
- [ ] privacy-indicator
- [ ] progressbar
- [ ] status-tag
- [ ] streamfield
- [ ] tabs
- [ ] tag
- [ ] tooltips

**Comment by jonnyscholes** [comment]
Excellent! Happy to start on any of these once #4979 is merged.

**Comment by thibaudcolas** [comment]
Thank you all for getting #4979 / #5215 finally merged ❤️.

With this major refactoring now in place, would now be a good time to introduce stricter linting rules? Or [Prettier](https://prettier.io/)?

Here are some of the linting rules I could think of as being useful:

- [`selector-max-type`](https://stylelint.io/user-guide/rules/selector-max-type/): `1`
- [`selector-no-qualifying-type`](https://stylelint.io/user-guide/rules/selector-no-qualifying-type/)
- [`selector-max-specificity`](https://stylelint.io/user-guide/rules/selector-max-specificity/): `0,3,0` (though to retrofit, would need plenty of ignore comments for the current code)
- [`selector-no-vendor-prefix`](https://stylelint.io/user-guide/rules/selector-no-vendor-prefix/)
- [`media-feature-name-no-vendor-prefix`](https://stylelint.io/user-guide/rules/media-feature-name-no-vendor-prefix/)
- [`at-rule-no-vendor-prefix`](https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix/)
- [`scss/declaration-nested-properties`](https://github.com/kristerkari/stylelint-scss#readme): `never`

These come from my [personal config](https://github.com/thibaudcolas/stylelint-config-cookbook), for which I reviewed all available rules about 6 months ago. There might be other ones available since then.

**Comment by nmorduch** [comment]
I've been considering whether it would be better to add stricter linting before or after refactoring each file. I'm feeling inclined at the moment toward before. I might suggest that for things like selector-max-specificity we can disable that rule for each file that has a problem, and then as each file is refactored, reenable it.

I don't have strong feelings at this point about what the rules are, just that there be rules, so I'm happy to defer to what you (and others) prefer.

**Comment by jonnyscholes** [comment]
@thibaudcolas I think the base Prettier config + those stylelint rules would be a brilliant place to start.

What if we turned on linting - but don't make it break the build - then over the next couple of releases work on removing all the warnings. Once thats done then flick the switch to force developers to adhere as a hard and fast rule?

**Comment by thibaudcolas** [comment]
👌 I'd be up for making a PR with new stylelint rules if you're ok to review that. Can also extract the config to a separate repo/package like we did for ESLint (https://github.com/wagtail/eslint-config-wagtail) if that seems valuable?

For the strategy to increase linting, my preference would be not to introduce any kind of project-wide "off time", as these can easily last longer than intended. I would see two options:

- Make linting more strict, and add ignore comments to legacy code that does not comply
- If stylelint supports that, have separate configs for legacy code and not-legacy (with config inheritance ideally)

In both cases it would probaby be better if the "base" config had our ideal config, and then legacy overrides made it less strict, so future developments know what baseline to aim for.

**Comment by jonnyscholes** [comment]
@thibaudcolas Yep I'd be down to review. 

Out of interest why was the ESLint config pulled out into a separate repo?

Yeah good point RE it just pushing out the inevitable. I feel like ignore comments is probably going to require the least effort and be simplest to setup quickly - and give us more flexibility without having to ad tweaks to the build config.

**Comment by thibaudcolas** [comment]
@jonnyscholes the general idea was that it would make it easier for us to reuse the config in other projects related to Wagtail, and have it available for people who would be developing something with the goal of eventually have it merged into Wagtail.

I think it somehow work(ed), but it was never advertised much, and it became out of date.

**Comment by jonnyscholes** [comment]
Ah nice, hadn't thought of that advantage. Yeah that makes sense - seems more and more projects are coming into the Wagtail org so may well get some use :)

**Comment by lb-** [comment]
@thibaudcolas  - I have recently worked through a revisit of our stylelint configs with a project I am working on. I think it is a great idea to separate out our ideal (most strict) config into its own repo, and then progressively use it across the scss as we work on it.

Stylelint Packages:
* https://github.com/hudochenkov/stylelint-order - adds rules for enforcing alphabetical ordering (eg. `background` before `color`), this may come across as a bit too strict but it really helps to minimise diffs, pick up obvious issues and work in scss that others are using - there may be alternatives though.
* https://github.com/hugomrdias/prettier-stylelint - CLI tool, see below

Styelint Config:
I had a look through our existing config, it looks pretty good, I think some rules may need to be re-checked that they are available still.

It would be good to extend the `stylelint-config-standard` as it is a good set of configs to start with - https://github.com/stylelint/stylelint-config-standard/blob/master/index.js

* `max-line-length` - this should match the prettier config default IMO - as such, we use 80 chars with `"ignore": ["comments", "non-comments"]` - https://stylelint.io/user-guide/rules/max-line-length/ 
* `media-feature-name-no-vendor-prefix` - + already mentioned - https://stylelint.io/user-guide/rules/media-feature-name-no-vendor-prefix/
* `order/properties-alphabetical-order` - this is from the stylelint-order plugin mentioned above
* `selector-no-vendor-prefix` - should not be needed with auto-prefixing + already mentioned - https://stylelint.io/user-guide/rules/selector-no-vendor-prefix/


Stylelint Tooling:
* We have used [Prettier Stylelint](https://github.com/hugomrdias/prettier-stylelint) it is a CLI that runs stylelint fix and then prettier over the top, it helps to easily enforce many of the stylelint rules in an automated way (it makes the properties ordering practical to use).
* Stylelint and prettier can work well together, they do different roles and I think we should use both for our scss files.

**Comment by nmorduch** [comment]
I just noticed that we've been discussing this on the general overhaul ticket, and am moving this conversation to a new ticket, #5246 :)

**Comment by jonnyscholes** [comment]
I've just noticed that the Explorer's scss resides outside of the main scss folder - this made sense when it was being developed as there was no existing place to put it. Should it be moved to `client/scss/components/`? (Same story for `.c-spinner`, `.c-status` and the `hallo` related scss). 

cc @nmorduch @thibaudcolas

**Comment by thibaudcolas** [comment]
I find it more valuable to separate code by component than by file type, especially for larger codebases like Wagtail’s. This works relatively well with React code with its component-driven nature, but if it was up to me I'd also extend this to HTML templates in Django codebases – it's quite annoying to have to hunt for templates when looking at given styles, and also the other way around.

**Comment by jonnyscholes** [comment]
@thibaudcolas Yeah I'd agree and would much prefer that too. But is that approach possible given how tied to the backend the HTML is and that only some parts of the frontend will ever be React? There has been chatter somewhere of turning everything into includes/macros - but thats a massive (and additional) job thats an even harder moving target because it directly impacts work done by the "backend team".

I guess in the mean time I'd prefer a consistent but no-quite-what-we-want approach than a mix of 2.

**Comment by jonnyscholes** [comment]
Also but I think there are complications when trying to use includes from outside the paths the template loader would search in? (Which anything inside `client` would be currently I think...)

**Comment by thibaudcolas** [comment]
As an additional task for this refactor, I’d like to propose we remove `home.css` and refactor most/all of its styles to components.

A good third of that file is overrides to the header, another third is the "X pages, Y images, Z documents" metrics, and the last third seems to be dead code ([search bar overrides](https://github.com/wagtail/wagtail/blame/master/wagtail/admin/static_src/wagtailadmin/scss/layouts/home.scss#L97-L123), I can’t see how a search bar could make its way to the homepage)

**Comment by nmorduch** [comment]
Closing this since we're going about this differently
